sdk/src/beta9/runner/function.py

```diff
@@ -131,87 +125,120 @@ def main(channel: Channel):
-            function_stub=function_stub,
-            gateway_stub=gateway_stub,
-        )
-        futures.thread._threads_queues.clear()
-
-        task_status = TaskStatus.Complete
-        error: Optional[BaseException] = None
 
-        # Invoke function
-        try:
-            get_args_resp: FunctionGetArgsResponse = function_stub.function_get_args(
-                FunctionGetArgsRequest(task_id=task_id)
+        # Start monitoring the task
+        with ThreadPoolExecutor() as thread_pool:
+            monitor_future = thread_pool.submit(
+                _monitor_task,
+                context=context,
+                stub_id=config.stub_id,
+                task_id=task_id,
+                container_id=container_id,
+                function_stub=function_stub,
+                gateway_stub=gateway_stub,
             )
 
-            if not get_args_resp.ok:
-                raise InvalidFunctionArgumentsException
-
-            payload: dict = _load_args(get_args_resp.args)
-            args = payload.get("args") or []
-            kwargs = payload.get("kwargs") or {}
-
-            handler = FunctionHandler()
-            result = handler(context, *args, **kwargs)
-        except BaseException as exc:
-            print(traceback.format_exc())
-            result = error = exc
-            task_status = TaskStatus.Error
-        finally:
-            pickled_result = cloudpickle.dumps(result)
-            set_result_resp: FunctionSetResultResponse = function_stub.function_set_result(
-                FunctionSetResultRequest(task_id=task_id, result=pickled_result)
+            # Invoke the function and handle its result
+            try:
+                result = invoke_function(function_stub, context, task_id)
+            except BaseException as e:
+                handle_task_failure(e, gateway_stub, task_id, container_id, container_hostname)
+                raise
+            finally:
+                monitor_future.cancel()
+
+            # End the task and send callback
+            complete_task(
+                gateway_stub,
+                result,
+                task_id,
+                container_id,
+                container_hostname,
+                start_time,
             )
-            if not set_result_resp.ok:
-                raise RunnerException("Unable to set function result")
 
-        task_duration = time.time() - start_time
 
-        end_task_response = end_task_and_send_callback(
-            gateway_stub=gateway_stub,
-            payload=result,
-            end_task_request=EndTaskRequest(
-                task_id=task_id,
-                task_duration=task_duration,
-                task_status=task_status,
-                container_id=container_id,
-                container_hostname=container_hostname,
-                keep_warm_seconds=0,
-            ),
-            override_callback_url=kwargs.get("callback_url"),
-        )
+def start_task(gateway_stub, task_id, container_id):
```

Comment:
did this functioanlly change anything?

Reply:
No, but it works with the theme of having the functions described as invoke, handle error, and complete task. It looks cleaner is all. Functionally, this does not worsen the code, it just makes it easier to read.

Reply:
makes sense